### DISCUSSION TITLE
File writes hanging (version 3.4.1) with Nano RP2040 Connect and flash memory chip from ISSI

### DISCUSSION THREAD

**Discussion by stefat** [khoih-prog/LittleFS_Mbed_RP2040#7] [Q&A] [answered]
**Settings:**

> Hw Mac Apple Silicon M1Pro
> macOS Ventura 13.0.1
> Arduino IDE 1.8.19
> Nano RP2040 Connect sdk core: Arduino Mbed OS Nano core 3.4.1

**Libraries:**

> LittleFS_Mbed_RP2040 3.4.1
> WiFiNINA_Generic version 1.8.14-7 

**The Issue:**

I'm using a Nano RP2040 Connect with **flash memory chip from ISSI**.
Trying to write several (17) files into the littleFS memory the system works fine for the first few files, say 3/4. How many files successfully written appears depending on their file sizes . 
After few files correctly written (verified using the readFile(<filepath>) method), always, the sw hangs.

I wonder if there is any workaround to overcome this issue **without** downgrading the core to version 2.3.1 as it brings other problems to my project.

**The code:**
The method involved is the following:

```
bool writeAllWebsiteDefaultFiles()
{
  bool success = true;
  bool filesuccess = true;
  size_t charsToRead;
  //### mystyle.css
  charsToRead = sizeof(css_mystyle);
  //deleteFile(path_mystyle);
  //usb.print(F("File size:"));usb.println(charsToRead);
  filesuccess = writeFile(path_mystyle, css_mystyle,charsToRead);
  success = filesuccess;
  //readFile(path_mystyle);
  //### index.htm
  charsToRead = sizeof(htm_index);
  //deleteFile(path_index);
  filesuccess = writeFile(path_index, htm_index,charsToRead);
  success = success && filesuccess;
  //readFile(path_index);
  //### info.htm
  charsToRead = sizeof(htm_info); 
  //deleteFile(path_info);
  filesuccess = writeFile(path_info, htm_info,charsToRead);
  success = success && filesuccess;
  //readFile(path_info);
  //### notfound.htm
  charsToRead = sizeof(htm_notfound);
  //deleteFile(path_notfound);
  filesuccess = writeFile(path_notfound, htm_notfound,charsToRead);
  //filesuccess ? usb.println(F("  notfound.htm ok")) : usb.println(F("  notfound.htm ko"));
  success = success && filesuccess;
  //### page1.htm
  charsToRead = sizeof(htm_page1); 
  //deleteFile(path_page1);
  filesuccess = writeFile(path_page1, htm_page1,charsToRead);
  //filesuccess ? usb.println(F("  page1.htm ok")) : usb.println(F("  page1.htm ko"));
  success = success && filesuccess;
  //readFile(path_page1);
  //### page2.htm
  charsToRead = sizeof(htm_page2);
  //deleteFile(path_page2);
  filesuccess = writeFile(path_page2, htm_page2,charsToRead);
  //filesuccess ? usb.println(F("  page2.htm ok")) : usb.println(F("  page2.htm ko"));
  success = success && filesuccess;
  //readFile(path_page2);

//###### ~10 times more the same code for more files

   success ? usb.println(F(">>> All files correctly written")) : usb.println(F(">>> Oops: error "));
  return success;
}
```

The above method run shows in the serial monitor:

```
10:06:00.068 ->   Writing file: /littlefs/mystyle.css>>> Open OK
10:06:00.220 -> >>> Writing OK
10:06:00.296 ->   Writing file: /littlefs/index.htm>>> Open OK
10:06:00.726 -> >>> Writing OK
10:06:00.794 ->   Writing file: /littlefs/info.htm>>> Open OK
10:06:00.945 -> >>> Writing OK
10:06:00.983 ->   Writing file: /littlefs/notfound.htm>>> Open OK
```

and then the system hangs.

Thank you for any idea on this.

**Comment by khoih-prog** [comment, marked as answer]
Hi @stefat 

This `ISSI` chip issue is well-known, discovered a very long time ago and still hasn't been solved.
You have to wait for Arduino to fix the core. **I'm also amazed that you can go so far**.

Check

1. [**RP2040 Connect board has faulty components in newest purchase** #318](https://github.com/arduino/ArduinoCore-mbed/issues/318)
2. [**WiFiManager_NINA_Lite-->RP2040_WiFiNINA Example does not work** #15](https://github.com/khoih-prog/WiFiManager_NINA_Lite/issues/15)

Try to use [**arduino-pico** core](https://github.com/earlephilhower/arduino-pico) to see if any better.

I have no way to figure out as I don't have that kind of special board ans also have no interest and time to investigate.

If it's a must, try using external flash-chip board, SD, etc., or even throw or return the board back to Arduino

Good Luck,

**Reply by stefat** [reply]
@khoih-prog Thank you. It's anyway important to me to know that. Appreciated.
No platform mbed/arduino-pico is perfect. I'm bouncing back and forth among them trying to fit my needs with their peculiar  features. I'll keep up trying and I'll keep you informed for any progress, just in case.I owe you.

**Reply by stefat** [reply]
Hi @khoih-prog 
For my application I found the PicoW with Arduino-pico core and Cypress wifi controller by far worse for WiFi stability than the Nina chip in the nano RP2040 connect.
I then returned back to the nano RP2040 connect with Arduino Mbed OS Nano core 3.4.1
1) as a workaround for the current discussion I found that your littleFS appears fully working with single files at a time. i.e. I removed in my app the writeAllWebsiteDefaultFiles() as in my first post and leave the user handle the FS. Upload/Save/Delete appears fully working so far. I did few successfully attempts: say ~30. No failure.
2) I've a question: I need to get the size of a file but the file.size() method is not found. The dirent struct also doesn't help. any clue ?

**Comment by khoih-prog** [comment]
You can try using the **Nano_RP2040_Connect** with `arduino-pico` core, which seems more supportive and stable.

**Reply by stefat** [reply]
it is, but at the time I'm on mbed that appears fine but for the OTA. I'll see. BTW: what about Littlefs file.size() method that I don't fin in mbed version ? Is my mistake or this method is not yet present in the library ?

**Comment by khoih-prog** [comment]
The current Mbed LittleFS implementation is using [**stdio.h**](https://www.tutorialspoint.com/c_standard_library/stdio_h.htm) of **C Library**. Therefore there is no **file.size()** function.

But you can use the combination of [**fseek**](https://www.tutorialspoint.com/c_standard_library/c_function_fseek.htm)

```cpp
int fseek(FILE *stream, long int offset, int whence);
```
with `offset = 0` and  `whence = SEEK_END` and 

function [**ftell**](https://www.tutorialspoint.com/c_standard_library/c_function_ftell.htm)

```cpp
long int ftell(FILE *stream);
```

to get the similar feature.

See the example

```cpp
#include <stdio.h>

int main ()
{
  FILE *fp;
  int len;

  fp = fopen("file.txt", "r");

  if ( fp == NULL )
  {
    perror ("Error opening file");
    return (-1);
  }

  fseek(fp, 0, SEEK_END);

  len = ftell(fp);
  fclose(fp);

  printf("Total size of file.txt = %d bytes\n", len);

  return (0);
}
```

**Reply by stefat** [reply]
it's all clear now. Thanks.
